pkg/ccl/crosscluster/logical/logical_replication_job_test.go

```diff
@@ -1031,6 +1031,155 @@ func TestHeartbeatCancel(t *testing.T) {
+
+	// Insert into dest, then check source2 -> dest wins
+	dbC.Exec(t, "UPSERT INTO tab VALUES (1, 'hello')")
+	dbB.Exec(t, "UPSERT INTO tab VALUES (1, 'goodbye')")
+	now := s.Clock().Now()
+	WaitUntilReplicatedTime(t, now, dbC, jobBID)
+	expectedRows := [][]string{
+		{"1", "goodbye"},
+	}
+	dbC.CheckQueryResults(t, "SELECT * from tab", expectedRows)
+
+	// Write to source1 and source2, which should keep their respective rows but dest should resolve a conflict
+	dbA.Exec(t, "UPSERT INTO tab VALUES (1, 'insertA')")
+	dbB.Exec(t, "UPSERT INTO tab VALUES (1, 'insertB')")
+
+	expectedRowsS1 := [][]string{
+		{"1", "insertA"},
+	}
+	expectedRowsDest := [][]string{
+		{"1", "insertB"},
+	}
+	now = s.Clock().Now()
+	WaitUntilReplicatedTime(t, now, dbC, jobAID)
+	WaitUntilReplicatedTime(t, now, dbC, jobBID)
+	dbA.CheckQueryResults(t, "SELECT * from tab", expectedRowsS1)
+	dbB.CheckQueryResults(t, "SELECT * from tab", expectedRowsDest)
+	dbC.CheckQueryResults(t, "SELECT * from tab", expectedRowsDest)
+}
+
+// TestFullyConnectedReplication tests 4 tables that are all streaming
+// from each other and how they handle conflicts
+func TestFullyConnectedReplication(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	skip.UnderRace(t, "Replication doesn't complete in time")
+
+	ctx := context.Background()
+
+	clusterArgs := base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			DefaultTestTenant: base.TestControlsTenantsExplicitly,
+			Knobs: base.TestingKnobs{
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			},
+		},
+	}
+
+	verifyExpectedRowAllServers := func(
+		t *testing.T, runners []*sqlutils.SQLRunner, expectedRows [][]string, dbNames []string,
+	) {
+		for i, name := range dbNames {
+			runners[i].CheckQueryResults(t, fmt.Sprintf("SELECT * from %s.tab", name), expectedRows)
+		}
+	}
+
+	waitUntilReplicatedTimeAllServers := func(
+		t *testing.T,
+		targetTime hlc.Timestamp,
+		runners []*sqlutils.SQLRunner,
+		jobIDs [][]jobspb.JobID,
+		numDBs int,
+	) {
+		for destDB := range numDBs {
+			for srcDB := range numDBs {
+				if destDB == srcDB {
+					continue
+				}
+				WaitUntilReplicatedTime(t, targetTime, runners[destDB], jobIDs[destDB][srcDB])
+			}
+		}
+	}
+
+	numDBs := 4
+	server, s, runners, dbNames := setupServerWithNumDBs(t, ctx, clusterArgs, 1, numDBs)
+	defer server.Stopper().Stop(ctx)
+
+	PGURLs, cleanup := GetPGURLs(t, s, dbNames)
+	defer cleanup()
+
+	// Each row is a DB, each column is a jobID from another DB to that target DB
+	jobIDs := make([][]jobspb.JobID, numDBs)
+	for i := range numDBs {
+		jobIDs[i] = make([]jobspb.JobID, numDBs)
+		for j := range numDBs {
+			if i == j {
+				jobIDs[i][j] = jobspb.InvalidJobID
+				continue
+			}
+			runners[i].QueryRow(t, "CREATE LOGICAL REPLICATION STREAM FROM TABLE tab ON $1 INTO TABLE tab", PGURLs[j].String()).Scan(&jobIDs[i][j])
```

Comment:
could you make this for loop more reader friendly by renaming j to `srcIdx` and i to `dstIdx`

Reply:
Fixed